src/main/java/org/truffleruby/language/arguments/EmptyArgumentsDescriptor.java

```diff
@@ -0,0 +1,20 @@
+/*
+ * Copyright (c) 2021 Oracle and/or its affiliates. All rights reserved. This
+ * code is released under a tri EPL/GPL/LGPL license. You can use it,
+ * redistribute it and/or modify it under the terms of the:
+ *
+ * Eclipse Public License version 2.0, or
+ * GNU General Public License version 2, or
+ * GNU Lesser General Public License version 2.1.
+ */
+package org.truffleruby.language.arguments;
+
+/** An arguments descriptor for where we're providing no information about the value of arguments being passed. */
```

Comment:
That sounds like `EmptyArgumentsDescriptor` can be used for all call sites, and receiving an empty EmptyArgumentsDescriptor doesn't imply there were no kwargs? I think that can't work, `EmptyArgumentsDescriptor` must mean no kwargs were passed.

Reply:
Currently it means no information was provided. When we change in subsequent PRs to add other constraints or guarantees I'll update the comment. But at the moment it's just the abscence of optional information.

Reply:
I feel this constraint is too weak. Either we give a proper descriptor in all places, or we're happy with just `EmptyArgumentsDescriptor` everywhere and add `NonEmptyArgumentsDescriptor` later.
A mix doesn't make sense to me, we'd try to figure the correct descriptor, but only in some arbitrary cases.

Reply:
In this PR, descriptors are optional, but tested, so we're happy to do whatever.

In a future PR, descriptors will be required to understand the keyword arguments, if the caller wants to send them flattened.

If the sender doesn't or can't send them flattened, for whatever reason, it can send them with an empty descriptor even though there's keyword arguments.

(May change to required as we fix the last of the Ruby 3 semantics.)

So I think the comment is correct now, and a mix is ok, but that may be tightened in the future, in which case I'll change the comment to match.

Reply:
> If the sender doesn't or can't send them flattened, for whatever reason, it can send them with an empty descriptor even though there's keyword arguments.
> (May change to required as we fix the last of the Ruby 3 semantics.)

How can one possibly know (in the callee) if it was a positional hash or kwargs passed to a method?
E.g. `def m(*args, **kwargs)` called with `m({a: 1})` vs `m(a: 1)` or potentially delegated multiple times through `(*a, **kw)`.
I believe the descriptor must track this (it could be a separate bit somewhere, but what's the point?), and `EmptyArgumentsDescriptor` must therefore means "no kwargs".